### DISCUSSION TITLE
Cannot find integration in HA

### DISCUSSION THREAD

**Discussion by ValiEne** [corvis/homeassistant_prana#2] [General]
Hi Dmitry, 
I've tried the HA integration and I'm facing some problems:
1. there are no files in releases section
2. I've downloaded and uploaded files into my HA instance and it looks like this:
**prana $ ls -alh**
total 60K    
drwxrwxrwx    3 root     root        4.0K Dec 18 18:39 .
drwxr-xr-x    5 root     root        4.0K Dec 22 23:11 ..
-rwxrwxrwx    1 root     root        3.5K Dec 18 18:39 __init__.py
-rwxrwxrwx    1 root     root        8.2K Dec 18 18:39 config_flow.py
-rwxrwxrwx    1 root     root        1.4K Dec 18 18:39 const.py
-rwxrwxrwx    1 root     root        2.2K Dec 18 18:39 entity.py
-rwxrwxrwx    1 root     root        8.0K Dec 18 18:39 fan.py
-rwxrwxrwx    1 root     root        1.8K Dec 18 18:39 light.py
-rwxrwxrwx    1 root     root         215 Dec 18 18:39 manifest.json
-rwxrwxrwx    1 root     root        2.6K Dec 18 18:39 switch.py
drwxrwxrwx    2 root     root        4.0K Dec 18 18:39 translations
-rwxrwxrwx    1 root     root         911 Dec 18 18:39 utils.py
**prana $ pwd**
/config/custom_components/prana
3. Restarted HA but I cannot find Prana Integration
What i'm missing?

**Comment by corvis** [comment]
Everything seem to be correct. Which HomeAssistant version do you have? I've tested this with 0.118 and there is a place in the code which relies on the pretty recent internal api, so I suspect it might be the problem... If that's the case I could rewrite the code to avoid new api call - shouldn't be hard.

**Reply by corvis** [reply]
anyway it would be great to see if you have any errors in the log during the startup.

**Comment by ValiEne** [comment]
I didn't find any reference about this Integration in logs. Any way to activate debug level of logging for this component?

I'm running the latest HASSIO, this are the details about my installation: 

version: 2020.12.1
installation_type: Home Assistant OS

host_os: Home Assistant OS 5.9
update_channel: stable
supervisor_version: 2020.12.7
docker_version: 19.03.13
Board: rpi4-64

**Reply by corvis** [reply]
Just updated my HASS to the recent version - everything seem to work fine. This makes me think it might be something specific to hass os and the way it manages custom components which prevents it from picking up this one... 
Do you have any other custom components? Do they work as expected?

In order to enable extra logs for prana add the following line to your logger config: `custom_components.prana: debug`
So you have something like:
```yaml
logger:
  default: info
  logs:
    custom_components.prana: debug
```

If you see no mentions of `prana` in logs it means HASS even doesn't try to load it. I expect to see at least these records in the log:
```
[homeassistant.loader] Loaded prana from custom_components.prana
[homeassistant.loader] You are using a custom integration for prana which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant.
[homeassistant.bootstrap] Domains to be set up: {'logger', 'device_automation', 'map', 'webhook', 'scene', 'frontend', 'http', 'media_source', 'zeroconf', 'prana', 'default_config', 'updater', 'blueprint', 'system_health', 'zone', 'alexa', 'met', ................

```

**Reply by corvis** [reply]
**UPDATE**: I tried to check on the fresh and clean installation of the recent hass. While I wan't able to reproduce your problem (missing integration in the integrations list) it revealed a serious dependency issue + a few config flow related issue which blocked me from completing setup.

I managed to fix all mentioned above which lead to a new release of prana_rc and multiple updates of homeassistant_prana so please update both. For homeassistant_prana check download the recent version from github and re-upload files.

**Comment by ValiEne** [comment]
Thanks Dmitry, somehow, after a number of reboots (HA and Hypervisor) the integration appeared at some point.
Now I need to see why a Rpi3 who was supposed to run prana-rc in http mode doesn't work (when I'm trying Integration I get an error "Unable to connect to Prana RC server"

**Reply by corvis** [reply]
Please check if prana rc docker image is running. Make sure you've set the correct port - if you didn't override the default docker settings the url for you should be like `http://YOUR-IP-PADDRESS:8881`.  BTW you could put it in the address line of your browser, if docker with prana is running you should see `405: Method Not Allowed`.

Also check the connectivity between your hass device and RPi3 with Prana RC. The easiest way would be just to call `curl http://orion.cnet.lan:8881/` from the hass device (not sure if HassOS allows ssh access and has curl).

And the last - check logs with `debug` logging activated for prana integration right after you get this error. The problem here is Home Assistant Config Flow api doesn't allow developer to render exact error message so you see pretty generic one while teh detailed message + stack trace will be in the log.

**Reply by ValiEne** [reply]
I'll try also the docker path in the following days. Now I just wrote a Raspbian image, booted a spare RPi3 and ran pip3 install prana-rc[server-tornado] command (I get an error if i'm trying to run pip instead of pip3). I'll keep you posted

**Reply by ValiEne** [reply]
Ok, I've managed to install Docker on Raspbian. Two observations: 
1. the docker command has to be: pi@raspberrypi:  sudo docker run --volume /run/dbus/system_bus_socket:/run/dbus/system_bus_socket --restart=unless-stopped **corvis/prana-rc:latest** 

2. even after running the prana-rc image I still cannot get an answer from the http server: 
pi@raspberrypi:~ $ curl   -X POST   -H "Accept: application/json"   http://192.168.11.79:8881/   -d '{ "jsonrpc": "2.0", "id": 1, "method": "prana.discover" }'
curl: (7) Failed to connect to 192.168.11.79 port 8881: Connection refused
pi@raspberrypi:~ $ curl   -X POST   -H "Accept: application/json"   http://localhost:8881/   -d '{ "jsonrpc": "2.0", "id": 1, "method": "prana.discover" }'
curl: (7) Failed to connect to localhost port 8881: Connection refused

**Reply by corvis** [reply]
You need to add a port mapping to make docker container port connected to the physical device port. Try `sudo docker run --volume /run/dbus/system_bus_socket:/run/dbus/system_bus_socket -p 8881:8881 --restart=unless-stopped corvis/prana-rc:latest`

**Reply by ValiEne** [reply]
You are right, I wasn't mapping the port 8881.
Now it works, I get this message when browsing to RPi IP address: <html><title>405: Method Not Allowed</title><body>405: Method Not Allowed</body></html>
With all of this the Integration refuses to work: 
![image](https://user-images.githubusercontent.com/47494711/103108267-f1059380-464e-11eb-9401-f489c79e4485.png)

**Reply by ValiEne** [reply]
Later edit: I'm using the latest versions for both prana-rc and HA Integration

**Reply by corvis** [reply]
Please show homeassistant logs right after the error

**Reply by ValiEne** [reply]
Here it is: 
`Logger: custom_components.prana.config_flow
Source: custom_components/prana/config_flow.py:44
Integration: Prana Recuperators
First occurred: 8:18:41 AM (1 occurrences)
Last logged: 8:18:41 AM

Unable to connect to prana rc server
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 975, in _wrap_create_connection
    return await self._loop.create_connection(*args, **kwargs)  # type: ignore  # noqa
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 1025, in create_connection
    raise exceptions[0]
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 1010, in create_connection
    sock = await self._connect_sock(
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 924, in _connect_sock
    await self.sock_connect(sock, address)
  File "/usr/local/lib/python3.8/asyncio/selector_events.py", line 496, in sock_connect
    return await fut
  File "/usr/local/lib/python3.8/asyncio/selector_events.py", line 528, in _sock_connect_cb
    raise OSError(err, f'Connect call failed {address}')
ConnectionRefusedError: [Errno 111] Connect call failed ('192.168.11.79', 8881)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/config/custom_components/prana/config_flow.py", line 44, in async_step_remote_connection
    await prana_client.healthcheck()
  File "/usr/local/lib/python3.8/site-packages/prana_rc/contrib/client/common.py", line 67, in healthcheck
    await self.async_invoke("prana.healthcheck", expected_response_type=PranaHealthCheckResultDTO),
  File "/usr/local/lib/python3.8/site-packages/sizzlews/client/common.py", line 179, in async_invoke
    return await self._invoke_request(rq, expected_response_type)
  File "/usr/local/lib/python3.8/site-packages/sizzlews/client/aiohttp.py", line 46, in _invoke_request
    response = await self._session.post(self.endpoint, json=rq.to_dict())
  File "/usr/local/lib/python3.8/site-packages/aiohttp/client.py", line 527, in _request
    conn = await self._connector.connect(
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 541, in connect
    proto = await self._create_connection(req, traces, timeout)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 898, in _create_connection
    _, proto = await self._create_direct_connection(req, traces, timeout)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 1057, in _create_direct_connection
    raise last_exc
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 1026, in _create_direct_connection
    transp, proto = await self._wrap_create_connection(
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 981, in _wrap_create_connection
    raise client_error(req.connection_key, exc) from exc
aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host 192.168.11.79:8881 ssl:default [Connect call failed ('192.168.11.79', 8881)]`

**Reply by corvis** [reply]
Hm no surprises here. The error really indicates inability to establish network connection. So you should double check if nothing blocks network communication. A few ideas to check:
- ensure docker run command is still runing
- ensure network connectivity between rpi3 and hass device
- ensure you don't have firewall blocking this connection on rpi3, hass, your router

**Reply by ValiEne** [reply]
It doesn't look like a connectivity problem. Here are some packets captured on the prana-rc server (192.168.11.71 is my laptop and .123 is the HA server):

**pi@RPi3B:~ $ sudo tcpdump port 8881 and host 192.168.11.123**
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
11:14:42.304768 IP 192.168.11.123.46602 > 192.168.11.79.8881: Flags [S], seq 1623826685, win 64240, options [mss 1460,sackOK,TS val 2371049689 ecr 0,nop,wscale 7], length 0
11:14:42.305116 IP 192.168.11.79.8881 > 192.168.11.123.46602: Flags [S.], seq 3929401319, ack 1623826686, win 65160, options [mss 1460,sackOK,TS val 2016197475 ecr 2371049689,nop,wscale 7], length 0
11:14:42.305603 IP 192.168.11.123.46602 > 192.168.11.79.8881: Flags [.], ack 1, win 502, options [nop,nop,TS val 2371049690 ecr 2016197475], length 0
11:14:42.313522 IP 192.168.11.123.46602 > 192.168.11.79.8881: Flags [P.], seq 1:181, ack 1, win 502, options [nop,nop,TS val 2371049698 ecr 2016197475], length 180
11:14:42.313791 IP 192.168.11.79.8881 > 192.168.11.123.46602: Flags [.], ack 181, win 508, options [nop,nop,TS val 2016197484 ecr 2371049698], length 0
11:14:42.314416 IP 192.168.11.123.46602 > 192.168.11.79.8881: Flags [P.], seq 181:253, ack 1, win 502, options [nop,nop,TS val 2371049699 ecr 2016197484], length 72
11:14:42.314576 IP 192.168.11.79.8881 > 192.168.11.123.46602: Flags [.], ack 253, win 508, options [nop,nop,TS val 2016197485 ecr 2371049699], length 0
^C
7 packets captured
12 packets received by filter
5 packets dropped by kernel

**pi@RPi3B:~ $ sudo tcpdump port 8881 and host 192.168.11.71**
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
11:16:48.718869 IP 192.168.11.71.64782 > 192.168.11.79.8881: Flags [F.], seq 1722775683, ack 3846512404, win 2051, options [nop,nop,TS val 501836517 ecr 1965604589], length 0
11:16:48.723283 IP 192.168.11.79.8881 > 192.168.11.71.64782: Flags [F.], seq 1, ack 1, win 503, options [nop,nop,TS val 1965613595 ecr 501836517], length 0
11:16:48.727405 IP 192.168.11.71.64782 > 192.168.11.79.8881: Flags [.], ack 2, win 2051, options [nop,nop,TS val 501836526 ecr 1965613595], length 0
11:16:51.468554 IP 192.168.11.71.64826 > 192.168.11.79.8881: Flags [S], seq 1192294227, win 65535, options [mss 1460,nop,wscale 6,nop,nop,TS val 501839248 ecr 0,sackOK,eol], length 0
11:16:51.468779 IP 192.168.11.71.64827 > 192.168.11.79.8881: Flags [S], seq 2006059860, win 65535, options [mss 1460,nop,wscale 6,nop,nop,TS val 501839248 ecr 0,sackOK,eol], length 0
11:16:51.469114 IP 192.168.11.79.8881 > 192.168.11.71.64826: Flags [S.], seq 4117803332, ack 1192294228, win 65160, options [mss 1460,sackOK,TS val 1965616340 ecr 501839248,nop,wscale 7], length 0
11:16:51.469278 IP 192.168.11.79.8881 > 192.168.11.71.64827: Flags [S.], seq 2361697279, ack 2006059861, win 65160, options [mss 1460,sackOK,TS val 1965616340 ecr 501839248,nop,wscale 7], length 0
11:16:51.475391 IP 192.168.11.71.64826 > 192.168.11.79.8881: Flags [.], ack 1, win 2058, options [nop,nop,TS val 501839253 ecr 1965616340], length 0
11:16:51.477622 IP 192.168.11.71.64827 > 192.168.11.79.8881: Flags [.], ack 1, win 2058, options [nop,nop,TS val 501839254 ecr 1965616340], length 0
11:16:51.483091 IP 192.168.11.71.64826 > 192.168.11.79.8881: Flags [P.], seq 1:477, ack 1, win 2058, options [nop,nop,TS val 501839256 ecr 1965616340], length 476
11:16:51.483357 IP 192.168.11.79.8881 > 192.168.11.71.64826: Flags [.], ack 477, win 506, options [nop,nop,TS val 1965616355 ecr 501839256], length 0
11:16:51.488751 IP 192.168.11.79.8881 > 192.168.11.71.64826: Flags [P.], seq 1:239, ack 477, win 506, options [nop,nop,TS val 1965616360 ecr 501839256], length 238
405 GET / (192.168.11.71) 3.70ms
11:16:51.493712 IP 192.168.11.71.64826 > 192.168.11.79.8881: Flags [.], ack 239, win 2055, options [nop,nop,TS val 501839272 ecr 1965616360], length 0

**Reply by corvis** [reply]
That's weird. When you click the Submit button on hass the only thing it does is sending `prana.healthcheck` to werify connectivity. Could you please keep tcpdump running and try to run this request from your laptop and then from hass so we could compare results:

```
curl \
  -X POST \
  -H "Accept: application/json" \
  http://YOUR-PRANA-RC-IP:8881/ \
  -d '{ "jsonrpc": "2.0", "id": 1, "method": "prana.healthcheck" }'
```

If you don't have curl on laptop you may use any other tcphttp client (e.g. postman) 
![Selection_149](https://user-images.githubusercontent.com/5593615/103135104-ad4a7280-46be-11eb-9ab1-1d563d23072a.png)

**Reply by ValiEne** [reply]
From MacOS: 
**curl \
   -X POST \
   -H "Accept: application/json" \
   http://192.168.11.79:8881/ \
   -d '{ "jsonrpc": "2.0", "id": 1, "method": "prana.healthcheck" }'**
{"error": {"code": -32601, "message": "Method not found"}, "id": 1, "jsonrpc": "2.0"}

TCPDump output: 
**sudo tcpdump port 8881 and host 192.168.11.71**
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
12:41:50.536004 IP 192.168.11.71.49589 > 192.168.11.79.8881: Flags [S], seq 1231505442, win 65535, options [mss 1460,nop,wscale 6,nop,nop,TS val 504730748 ecr 0,sackOK,eol], length 0
12:41:50.536556 IP 192.168.11.79.8881 > 192.168.11.71.49589: Flags [S.], seq 4230028523, ack 1231505443, win 65160, options [mss 1460,sackOK,TS val 1970715408 ecr 504730748,nop,wscale 7], length 0
12:41:50.544581 IP 192.168.11.71.49589 > 192.168.11.79.8881: Flags [.], ack 1, win 2058, options [nop,nop,TS val 504730757 ecr 1970715408], length 0
12:41:50.544703 IP 192.168.11.71.49589 > 192.168.11.79.8881: Flags [P.], seq 1:226, ack 1, win 2058, options [nop,nop,TS val 504730757 ecr 1970715408], length 225
12:41:50.544999 IP 192.168.11.79.8881 > 192.168.11.71.49589: Flags [.], ack 226, win 508, options [nop,nop,TS val 1970715416 ecr 504730757], length 0
12:41:50.552713 IP 192.168.11.79.8881 > 192.168.11.71.49589: Flags [P.], seq 1:221, ack 226, win 508, options [nop,nop,TS val 1970715424 ecr 504730757], length 220
12:41:50.556862 IP 192.168.11.71.49589 > 192.168.11.79.8881: Flags [.], ack 221, win 2055, options [nop,nop,TS val 504730772 ecr 1970715424], length 0

**Reply by ValiEne** [reply]
From HassOS: 
**curl \
   -X POST \
   -H "Accept: application/json" \
   http://192.168.11.79:8881/ \
   -d '{ "jsonrpc": "2.0", "id": 1, "method": "prana.healthcheck" }'**
{"error": {"code": -32601, "message": "Method not found"}, "id": 1, "jsonrpc": "2.0"}

TCPDump: 
**sudo tcpdump port 8881 and host 192.168.11.123**
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
12:45:38.680736 IP 192.168.11.123.50514 > 192.168.11.79.8881: Flags [S], seq 1037262531, win 64240, options [mss 1460,sackOK,TS val 4123196620 ecr 0,nop,wscale 7], length 0
12:45:38.681109 IP 192.168.11.79.8881 > 192.168.11.123.50514: Flags [S.], seq 1161737637, ack 1037262532, win 65160, options [mss 1460,sackOK,TS val 2021653851 ecr 4123196620,nop,wscale 7], length 0
12:45:38.681619 IP 192.168.11.123.50514 > 192.168.11.79.8881: Flags [.], ack 1, win 502, options [nop,nop,TS val 4123196621 ecr 2021653851], length 0
12:45:38.681698 IP 192.168.11.123.50514 > 192.168.11.79.8881: Flags [P.], seq 1:226, ack 1, win 502, options [nop,nop,TS val 4123196621 ecr 2021653851], length 225
12:45:38.681980 IP 192.168.11.79.8881 > 192.168.11.123.50514: Flags [.], ack 226, win 508, options [nop,nop,TS val 2021653852 ecr 4123196621], length 0
12:45:38.689904 IP 192.168.11.79.8881 > 192.168.11.123.50514: Flags [P.], seq 1:221, ack 226, win 508, options [nop,nop,TS val 2021653860 ecr 4123196621], length 220
12:45:38.690383 IP 192.168.11.123.50514 > 192.168.11.79.8881: Flags [.], ack 221, win 501, options [nop,nop,TS val 4123196630 ecr 2021653860], length 0

**Reply by corvis** [reply]
oh. Now i see. You are runing Prana RC version 0.3.6. And the reason for that is because of some issue i  github build pipeline the recent docker image was never pushed to registry. I'll check and let you know when you could update prana rc

**Reply by ValiEne** [reply]
Good, I'm glad you saw the problem 👍

**Reply by corvis** [reply]
Pushed the new docker image. Could you please try again after updating docker image with this image `corvis/prana-rc:0.3.11` instead of `latest`. If still doesn't work please post hass logs and repeat the latest curl experiment.

**Reply by ValiEne** [reply]
Thanks a lot Dmitry, it works like a charm. Thanks a lot for your time, Merry Christmas

**Reply by corvis** [reply]
Oh wow, finally we've made it. Enjoy your new home automation! Also, please note - id doesn't support managing input and output flows separately at the moment as well as adjusting display brightness.

I'll update a doc to add port mapping thing which might confuse other people. Also, could you check if everything recovers when you reboot RPI3 with docker without your intervention? If not you might need to create a systemd module to run start container on boot. 

Have a wonderful Christmas too.

**Reply by corvis** [reply]
One more thing :) You might want to disable DEBUG logging for prana integration in your config as it generates a new line each 30 secs.

**Reply by ValiEne** [reply]
Hi Dmitry, 
Thanks again for your help, I really appreciate it.
No, after a reboot the image doesn't start automatically. 
I saw the Input/Output part and I didn't know what they are referring to. I guess your Pranas are newer then mine which is 3-4 years old and on my model the ingress/egress air flow are working in parallel, no way to make it work sequential, like 30 sec. bring clean air and next 30 sec. remove air from the room. I saw a similar system at NovingAir but they are do it with multiple fans, minimum two, with the units working in sync, one is extracting air, the other bringing air in thee room for few tens of seconds and then the flow changes direction.

**Comment by ValiEne** [comment]
I said you tried with a fresh install, using the latest available version. Did you used HassOS or other method?

**Reply by corvis** [reply]
No, I just used the recent official hass docker image and installed on a fresh linux VM. I never tried HassOS so have no idea how it works.

**Comment by corvis** [comment]
Btw, yesterday I finally managed to integrate it into my real home setup and it seems to work fine. 

![Selection_147](https://user-images.githubusercontent.com/5593615/103106316-1be6ec00-463d-11eb-94bf-8ef36e3726dc.png)

So I've put a RPi running prana-rc in the center of the house and connected it with hass server siting in a sever room by wire. 
Prana RC constantly keeps open bluetooth connections with my 3 Pranas which make the latency for turning on and of less then a second. So far so good, but we will see how stable this setup is...

**Reply by ValiEne** [reply]
Nice dashboard 👍
